### PR TITLE
Initial check-in for ID for DNS

### PR DESCRIPTION
**Pull request by aeijdenberg**
It's possibly not quite in a great condition to check-in, but I may be out for a few days so wanted to get something out.

### CODE REVIEW

File: dns/draft-ct-over-dns.md [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+  RFC6962:
+  RFC1035:
+  RFC4648:
+
+informative:
+  RFC1034:
+
+--- abstract
+
+We describe a method for fetching certificate transparency inclusion proofs over recursive DNS.
```
Comment (benlaurie) [inline dns/draft-ct-over-dns.md]: Is it certificate transparency or Certificate Transparency?

File: dns/draft-ct-over-dns.md [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+: `leaf_hash` is the a leaf hash as defined in section 3.4 of RFC6962.
+
+Outputs:
+: 
+: `leaf_index` is the index of the `leaf_hash` in the Merkle tree.
+
+
+To query for this message, set `QNAME` to `<encoded_leaf_hash> || '.hash.' || <domain_for_log> || '.'`
+
+`encoded_leaf_hash` is the base32 encoding of `leaf_hash`.  For example, given a `MerkleTreeLeaf` structure represented as the concentation of:
```
Comment (benlaurie) [inline dns/draft-ct-over-dns.md]: Examples are usually given in appendices at the end.

File: dns/draft-ct-over-dns.md [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    0x0001020304050607 (timestamp)
+    0x0000             (x509_entry)
+    0x000001           (length of ASN.1 Cert)
+    0x00               (bogus ASN.1 Cert)
+    0x0000             (extensions length)
+
+
+Then the `encoded_leaf_hash` would be:
+
+    D4S6DSV2J743QJZEQMH4UYHEYK7KRQ5JIQOCPMFUHZVJNFGHXACA====
```
Comment (benlaurie) [inline dns/draft-ct-over-dns.md]: = is not a legal domain name character.

File: dns/draft-ct-over-dns.md [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+Since the length is known, a client MAY omit the 4 padding byte suffix.
```
Comment (benlaurie) [inline dns/draft-ct-over-dns.md]: Since = is illegal, a client MUST omit it.

File: dns/draft-ct-over-dns.md (hunk 2 above)
Comment (benlaurie) [inline dns/draft-ct-over-dns.md]: Since its integral to your narrative, leave it here.

File: dns/draft-ct-over-dns.md [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+: `domain_for_log` is a value stored in log metadata.  For example, a name server operated by Google for the Pilot log may have the domain `pilot.ct.googleapis.com`.
+
+: `leaf_hash` is the a leaf hash as defined in section 3.4 of RFC6962.
+
+Outputs:
+: 
+: `leaf_index` is the index of the `leaf_hash` in the Merkle tree.
+
+
+To query for this message, set `QNAME` to `<encoded_leaf_hash> || '.hash.' || <domain_for_log> || '.'`
```
Comment (benlaurie) [inline dns/draft-ct-over-dns.md]: You'll need to specify somewhere that `||` denote concatenation.

Or perhaps just say `<encoded leaf hash>.hash.<domain for log>.`

File: dns/draft-ct-over-dns.md [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+--- abstract
+
+We describe a method for fetching certificate transparency inclusion proofs over recursive DNS.
+
+--- middle
+
+# Introduction
+
+Certificate Transparency improves the certificate ecosystem by requiring certificates be submitted to logs in order to be considered secure.  To prove that a certificate has been included in a public log clients should request inclusion proofs from logs as descrbed in RFC6962.
```
Comment (benlaurie) [inline dns/draft-ct-over-dns.md]: Need a reference for RFC6962.

Comment (benlaurie) [inline dns/draft-ct-over-dns.md]: Also, `RFC 6962`.

File: dns/draft-ct-over-dns.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+However by fetching an inclusion proof directly from a log server the client reveals to the log server the certificate that they are interested in, and thus leaks their browsing history.
+
+This document describes the use of special DNS records so that a client can look up certificate transparency information via their existing DNS resolver, meaning that they will identify their browsing history to only their existing DNS resolver (typically at their ISP), which, in most cases, already knows which sites the user has visited by virtue of having resolved their DNS requests already.
+
+In this manner only a proxy for aggregate usage of sites contained in a log will be detectable by the log operator or mirror that operates the authoritative name server used by these resolvers.
+
+Google plans to operate such a DNS resolver to put in front of log mirrors for at least all logs included by Chrome.  Other clients are welcome to use the same name servers if desired.
```
Comment (benlaurie) [inline dns/draft-ct-over-dns.md]: Inappropriate for an RFC. We can say that elsewhere.

File: dns/draft-ct-over-dns.md [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+The tree query takes as input:
+
+- `domain_for_log`
+- `start_index`
+- `leaf_index`
+- `tree_size`
+
+And returns as output:
+
+- `partial_audit_path[]`
```
Comment (benlaurie) [inline dns/draft-ct-over-dns.md]: Again you'll need to define that `[]` denotes an array.

File: dns/draft-ct-over-dns.md [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+               | <character-string for BASE64(audit_path[2])>      |
+               | ...                                               |
+               +---------------------------------------------------+
+    Authority  | <empty>                                           |
+               +---------------------------------------------------+
+    Additional | <empty>                                           |
+               +---------------------------------------------------+
+
+Where the TXT RDATA returned will be one or more character-strings, each containing the BASE64 encoding of the value for next value in the `audit_path` array.  Note that the first character-string will be for the number specified in the query, and the server MAY return one or more subsequent values that correspond to the next values in the array.  Since the size of DNS responses is limited, servers MAY not return all values for `audit_path` in a single query, and if the correct number (`length`) of response is not returned, a client MUST send a subsequent query with a different starting index until all elements are accounted for.
```
Comment (benlaurie) [inline dns/draft-ct-over-dns.md]: TXT RDATA is a single character-string containing as many b64 encoded entries as will fit (i.e. up to 7 for SHA-256).

Comment (pphaneuf) [inline dns/draft-ct-over-dns.md]: b64 encoded? Raw binary? 7 hashes of 32 bytes, once b64 encoded, that would be 299 characters, no?

File: dns/draft-ct-over-dns.md (hunk 1 above)
Comment (aeijdenberg) [inline dns/draft-ct-over-dns.md]: Capitals!  Have updated.

File: dns/draft-ct-over-dns.md (hunk 3 above)
Comment (aeijdenberg) [inline dns/draft-ct-over-dns.md]: Thanks, clarified.

File: dns/draft-ct-over-dns.md (hunk 4 above)
Comment (aeijdenberg) [inline dns/draft-ct-over-dns.md]: Done.

File: dns/draft-ct-over-dns.md (hunk 5 above)
Comment (aeijdenberg) [inline dns/draft-ct-over-dns.md]: Added explanation of ||.

File: dns/draft-ct-over-dns.md (hunk 6 above)
Comment (aeijdenberg) [inline dns/draft-ct-over-dns.md]: Done (I think).

File: dns/draft-ct-over-dns.md (hunk 7 above)
Comment (aeijdenberg) [inline dns/draft-ct-over-dns.md]: I assume you just mean the last paragraph?  Removing.

File: dns/draft-ct-over-dns.md (hunk 8 above)
Comment (aeijdenberg) [inline dns/draft-ct-over-dns.md]: Changed to same style as 6962.

File: dns/draft-ct-over-dns.md (hunk 9 above)
Comment (aeijdenberg) [inline dns/draft-ct-over-dns.md]: I think we settled on binary.  I'll update text to reflect.

**Comment by aeijdenberg** [comment]
Added a new commit with substantial revisions.  PTAL.

**Comment by aeijdenberg** [comment]
Gentle ping.  I think I've addressed the existing concerns.

File: dns/draft-ct-over-dns.md [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+
+--- middle
+
+# Introduction
+
+Certificate Transparency improves the certificate ecosystem by requiring certificates be submitted to logs in order to be considered secure.  To prove that a certificate has been included in a public log clients should request inclusion proofs from logs as descrbed in {{RFC6962}}.
+
+However by fetching an inclusion proof directly from a log server the client reveals to the log server the certificate that they are interested in, and thus leaks their browsing history.
+
+This document describes the use of special DNS records so that a client can look up Certificate Transparency information via their existing DNS resolver, meaning that they will identify their browsing history to only their existing DNS resolver (typically at their ISP), which, in most cases, already knows which sites the user has visited by virtue of having resolved their DNS requests already.
```
Comment (pphaneuf) [inline dns/draft-ct-over-dns.md]: Special records? They're just plain old records, really... :smiley:

File: dns/draft-ct-over-dns.md [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+
+--- abstract
+
+We describe a method for fetching Certificate Transparency inclusion proofs over recursive DNS.
+
+--- middle
+
+# Introduction
+
+Certificate Transparency improves the certificate ecosystem by requiring certificates be submitted to logs in order to be considered secure.  To prove that a certificate has been included in a public log clients should request inclusion proofs from logs as descrbed in {{RFC6962}}.
```
Comment (pphaneuf) [inline dns/draft-ct-over-dns.md]: s/descrbed/described/

Comment (aeijdenberg) [inline dns/draft-ct-over-dns.md]: Fixed, thanks.

File: dns/draft-ct-over-dns.md (hunk 10 above)
Comment (aeijdenberg) [inline dns/draft-ct-over-dns.md]: I think they're special.  If they weren't special, we wouldn't need this I-D.

Comment (pphaneuf) [inline dns/draft-ct-over-dns.md]: Some protocols define their own RRs, that's what I would consider "special".

Comment (aeijdenberg) [inline dns/draft-ct-over-dns.md]: Can you suggest a rephrase?

**Pull request merged**